### ISSUE TITLE
`NoReturn` as return type of overloaded function is not handled properly

### ISSUE THREAD

**Issue by Hackrrr** [microsoft/pyright#3800]
**Describe the bug**
Pyright doesn't seem to handle the `NoReturn` type correctly if a `NoReturn` is returned from an overloaded function:

```python
from typing import overload, NoReturn

@overload
def noreturn_overload() -> NoReturn: ...
@overload
def noreturn_overload(x: int) -> None: ...
def noreturn_overload(x: int = 0) -> NoReturn | None: ...

def f() -> NoReturn: # Error: Function with declared type of "NoReturn" cannot return "None"
    noreturn_overload() # reveal_type() output is "NoReturn"
```

**Expected behavior**
Pyright shouldn't report an error for above code.

**VS Code extension or command-line**
pyright cli 1.1.265

**Comment by erictraut** [comment]
This is a known limitation in the way that pyright implements `NoReturn` call evaluations. Such evaluations need to be performed during code flow analysis because they affect reachability within the code flow graph, but the code flow graph itself affects type evaluation. This circular dependency (where type analysis depends on the code flow graph, but the code flow graph depends on type analysis) is tricky to handle, and pyright makes some simplifying assumptions that are unable to handle certain edge cases like the one in this code sample. In particular, it doesn't fully evaluate overloaded functions or methods. Instead, it determines whether some (but not all) of the overloads return `NoReturn` and assumes that it's not a `NoReturn` call in that case.

I'd like to understand your use can better. Can you describe why you want to use `NoReturn` in an overload?

I think I can find a way to eliminate this limitation, but it might result in a (potentially significant) slowdown for type analysis. If I can limit the slowdown to this specific case, I think it might be worth considering. If it slows down analysis in all cases, then I don't think it's worth it.

**Comment by Hackrrr** [comment]
I'm building an application when I need to time function calls to certain "points in time" and I wrote class `Sync` which helps with this (little bit simplied from original):
```python
from datetime import datetime, timedelta
from time import sleep
from threading import Event
from typing import Callable, NoReturn, overload

class Sync:
    def __init__(self, sync_point: datetime, cycle: timedelta):
        self.sync_point: datetime = sync_point
        self.cycle: timedelta = cycle

    @property
    def to_next_cycle(self) -> float:
        # Some calculations
        ...

    @overload
    def run(self, func: Callable, stop_event: None = None) -> NoReturn: ...
    @overload
    def run(self, func: Callable, stop_event: Event | None = None) -> None: ...
    def run(self, func: Callable, stop_event: Event | None = None) -> NoReturn | None:
        time2wait: float
        while True:
            time2wait = self.to_next_cycle
            if stop_event is None:
                sleep(time2wait)
            else:
                if stop_event.wait(time2wait):
                    # Event was set, exiting
                    return
            func()
```
Sometimes I will keep these function running for eternity and sometimes I need a way to stop them - this is why I want to do overload with `NoReturn`. It's not really a necessity for me to have `NoReturn` there but I'm glad that I had it in there in previous iterations (when there wasn't need to stop these functions) as IDE (VS Code) warned me about unreachable code after calling `.run()` (because I really didn't realize that, even though it's obvious).

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.266, which I just published. It will also be included in a future release of pylance.
